lib/gossip/damper.js

```diff
@@ -0,0 +1,580 @@
+    if (!this.hasFlapper(member)) {
+        this.logger.warn('ringpop damper cannot damp member; member is not flappy', {
+            local: this.ringpop.whoami(),
+            member: addr
+        });
+        return false;
+    }
+
+    this.ringpop.membership.makeDamped(member.address);
+
+    // Don't start anymore damping subprotocols for a member
+    // that is already damped.
+    this.removeFlapper(member.address);
+    this.addDampedMember(member.address);
+
+    if (!this.isExpirationTimerEnabled) {
+        this.logger.debug('ringpop damper starting expiration timer; a member has been damped', {
+            local: this.ringpop.whoami(),
+            member: member.address
+        });
+        this._startExpirationTimer();
+    }
+
+    return true;
+};
+
+Damper.prototype.expireDampedMembers = function expireDampedMembers() {
+    var undampedMembers = [];
+    var suppressDuration = this.ringpop.config.get('dampScoringSuppressDuration');
+    var keys = Object.keys(this.dampedMembers);
+    for (var i = 0; i < keys.length; i++) {
+        var key = keys[i];
+        var memberDamping = this.dampedMembers[key];
+        var dampedDuration = this.Date.now() - memberDamping.timestamp;
+        if (dampedDuration >= suppressDuration) {
+            delete this.dampedMembers[key];
+            undampedMembers.push({
+                member: key,
+                timeDamped: dampedDuration
+            });
+        }
+    }
+
+    if (undampedMembers.length === 0) {
+        return undampedMembers;
+    }
+
+    this.emit('event', new events.DampedMemberExpirationEvent(undampedMembers));
+    this.logger.info('ringpop damper undamped members', {
+        local: this.ringpop.whoami(),
+        suppressDuration: suppressDuration,
+        undampedMembers: undampedMembers
+    });
+
+    // If there are no more damped members, stop trying to expire them.
+    if (Object.keys(this.dampedMembers).length === 0) {
+        this.logger.debug('ringpop damper stopping expiration timer; no damped members left', {
+            local: this.ringpop.whoami()
+        });
+        this._stopExpirationTimer();
+    }
+
+    return undampedMembers;
+};
+
+Damper.prototype.destroy = function destroy() {
+    this._stopDampTimer();
+    this._stopExpirationTimer();
+};
+
+Damper.prototype.isDamped = function isDamped(flapper) {
+    return !!this.dampedMembers[flapper.address];
+};
+
+Damper.prototype.hasFlapper = function hasFlapper(flapper) {
+    return !!this.flappers[flapper.address];
+};
+
+Damper.prototype.hasStarted = function hasStarted() {
+    return !!this.dampTimer;
+};
+
+Damper.prototype.initiateSubprotocol = function initiateSubprotocol(callback) {
+    var self = this;
+    var config = this.ringpop.config;
+
+    // TODO Stop damp timer when cluster has reached damped limits
+    if (!this._validateDampedClusterLimits()) {
+        callback();
+        return;
+    }
+
+    // You may think that a check to verify that flappers actually
```

Comment:
This comment felt out of left field for me, and I realized afterward that it's because "initiateSubprotocol" doesn't tell me intuitively what this function does. If it was `initiateFanout(Subprotocol)`, maybe I would have had a clear notion of what's going on. Or a high level comment explaining it perhaps.